server/src/main/java/org/opensearch/index/shard/CheckpointRefreshListener.java

```diff
@@ -40,7 +40,10 @@ public void beforeRefresh() throws IOException {
 
     @Override
     public void afterRefresh(boolean didRefresh) throws IOException {
-        if (didRefresh && shard.state() != IndexShardState.CLOSED && shard.getReplicationTracker().isPrimaryMode()) {
+        if (didRefresh
+            && shard.state() != IndexShardState.CLOSED
+            && shard.getReplicationTracker().isPrimaryMode()
+            && shard.isBlockInternalCheckPointRefresh() == false) {
```

Comment:
Why is primary mode not sufficient for publishing?

Lets assume the condition is correct, how do we ensure that the publisher stops lets say in usecase where once publish starts and then this condition becomes false? The primary term validation is present only for certain TransportReplicationActions. Is it possible that a particular target node could be receiving segment files from old primary as well a new primary given the afterRefresh method is async. If this understanding is correct, pls do help with the concerned code that handles this.

Reply:
Thats correct primary term bump does not happen during peer primary recoveries

Reply:
I think it can be handled in `clusterChanged` event inside [SegmentReplicationSourceService](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java#L159)

Reply:
@Bukhtawar : Tried the changes on local machine and I see seg rep took additional ~3.5 secs during relocation handoff. [More details](https://github.com/opensearch-project/OpenSearch/issues/5242#issuecomment-1368133952)

Reply:
Discussed this offline with @mch2, will take this up in a separate task. Created https://github.com/opensearch-project/OpenSearch/issues/5706 to track this separately.